pilot/pkg/simulation/traffic.go

```diff
@@ -0,0 +1,438 @@
+// The implementation may initially be confusing because of a property of the
+// Envoy algorithm - at each level we will filter out all FilterChains that do
+// not match. This means an empty match (`{}`) may not match if another chain
+// matches one criteria but not another.
+func (sim *Simulation) matchFilterChain(chains []*listener.FilterChain, input Call) (*listener.FilterChain, error) {
+	chains = filter(chains, func(fc *listener.FilterChainMatch) bool {
+		return fc.GetDestinationPort() == nil
+	}, func(fc *listener.FilterChainMatch) bool {
+		return int(fc.GetDestinationPort().GetValue()) == input.Port
+	})
+	chains = filter(chains, func(fc *listener.FilterChainMatch) bool {
+		return fc.GetPrefixRanges() == nil
+	}, func(fc *listener.FilterChainMatch) bool {
+		ranger := cidranger.NewPCTrieRanger()
+		for _, a := range fc.GetPrefixRanges() {
+			_, cidr, err := net.ParseCIDR(fmt.Sprintf("%s/%d", a.AddressPrefix, a.GetPrefixLen().GetValue()))
+			if err != nil {
+				sim.t.Fatal(err)
+			}
+			if err := ranger.Insert(cidranger.NewBasicRangerEntry(*cidr)); err != nil {
+				sim.t.Fatal(err)
+			}
+		}
+		f, err := ranger.Contains(net.ParseIP(input.Address))
+		if err != nil {
+			sim.t.Fatal(err)
+		}
+		return f
+	})
+	chains = filter(chains, func(fc *listener.FilterChainMatch) bool {
+		return fc.GetServerNames() == nil
+	}, func(fc *listener.FilterChainMatch) bool {
+		sni := host.Name(input.Sni)
+		for _, s := range fc.GetServerNames() {
+			if sni.SubsetOf(host.Name(s)) {
+				return true
+			}
+		}
+		return false
+	})
+	chains = filter(chains, func(fc *listener.FilterChainMatch) bool {
+		return fc.GetTransportProtocol() == ""
+	}, func(fc *listener.FilterChainMatch) bool {
+		switch fc.GetTransportProtocol() {
+		case xdsfilters.TLSTransportProtocol:
+			return input.Protocol == HTTPS || input.Protocol == TLS
+		case xdsfilters.RawBufferTransportProtocol:
+			return input.Protocol != HTTPS && input.Protocol != TLS
+		}
+		return false
+	})
+	chains = filter(chains, func(fc *listener.FilterChainMatch) bool {
+		return fc.GetApplicationProtocols() == nil
+	}, func(fc *listener.FilterChainMatch) bool {
+		return sets.NewSet(fc.GetApplicationProtocols()...).Contains(input.Alpn)
+	})
+	// We do not implement the "source" based filters as we do not use them
+	if len(chains) > 1 {
+		return nil, ErrMultipleFilterChain
+	}
+	if len(chains) == 0 {
+		return nil, ErrNoFilterChain
+	}
+	return chains[0], nil
+}
+
+func filter(chains []*listener.FilterChain,
+	empty func(fc *listener.FilterChainMatch) bool,
+	match func(fc *listener.FilterChainMatch) bool) []*listener.FilterChain {
+	res := []*listener.FilterChain{}
+	anySet := false
+	for _, c := range chains {
+		if !empty(c.GetFilterChainMatch()) {
+			anySet = true
+		}
+	}
+	if !anySet {
+		return chains
+	}
+	for _, c := range chains {
+		if match(c.GetFilterChainMatch()) {
+			res = append(res, c)
+		}
+	}
+	return res
+}
+
+func protocolToAlpn(s Protocol) string {
```

Comment:
so dont we have to consider all the other alpns that we shove into fcm? i know that this piece is for http sniffing but a lot of our filter chain matches have the metadata exchange related alpns and associated logic. Might be worth adding a TODO

Reply:
Those ALPNs really only impact inbound I think. This doesn't really handle inbound right now. I think inbound is lower priority - will add a todo